### PR TITLE
bpo-35936: Updates to modulefinder

### PR DESCRIPTION
**Pull request by brandtbucher**
This patch includes a few useful fixes to `Lib/modulefinder.py`. Namely:

- It doesn't crash if it encounters a syntax error ([bpo-17396](https://bugs.python.org/issue17396)).
- It doesn't report certain name collisions as bad ([bpo-35376](https://bugs.python.org/issue35376)).
- It doesn't use mutable default arguments in its initializer.
- Most importantly, it now uses `importlib` instead of `imp`, which is deprecated ([bpo-25160](https://bugs.python.org/issue25160), [bpo-20020](https://bugs.python.org/issue20020)) As a benefit, frozen built-in modules (`zipimport`, for example) are now correctly reported.

With the exception of these bug fixes, I've been very careful to preserve the original behavior, and have not changed any part of the API. This patch also includes 2 new regression tests.

https://bugs.python.org/issue35936

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2019-02-08-04-46-58.bpo-35936.q2ArjU.rst
Comment (pganssle) [inline Misc/NEWS.d/next/Library/2019-02-08-04-46-58.bpo-35936.q2ArjU.rst]: I personally do not like it when a changelog makes references to bugfixes without specifying what they are. I think maybe it would be appropriate to add separate news entries for each of the long-standing bugs that were fixed. I don't think there's any rule that says that a given PR must have only one changelog entry, and presumably it would be nice to have one changelog ticket per BPO issue number.

**Comment by ronaldoussoren** [comment]
This changeset makes a substantial change to the semantics of modulefinder: using ``importlib.util.find_spec`` can change the import state of the script. In particular ``find_spec("package.module")`` results in the importing of ``package`` (and adds it to ``sys.modules``).

Furthermore using ``importlib`` makes it close to impossible to use an alternate search path (the "path" argument for the constructor) due to the API of importlib. There are workarounds for this, but that probably requires replacing the entire import state (not just ``sys.path`` as this pull request does) around calls to importlib functions. 

P.S. For modulegraph2, which is a rewrite of modulegraph which itself is a library that uses code heavily inspired by modulefinder I choosen to take away the possibility to use an alternate search path.

**Comment by brandtbucher** [comment]
@ronaldoussoren I see, thanks for the input. As a concrete example, we can see import side-effects of the parent module when running `$ python -c 'import importlib; importlib.util.find_spec("this.that")'`. I'll dig into workarounds a bit more today. If nothing seems viable, as you mentioned, I'll probably revert those changes and leave the other fixes for merging.

It may be worth exploring a ground-up redesign, since the documented API is so small. I need a side-project, so I'd be willing to work on that if others feel it would be valuable.

File: Misc/NEWS.d/next/Library/2019-02-08-04-46-58.bpo-35936.q2ArjU.rst
Comment (brandtbucher) [reply Misc/NEWS.d/next/Library/2019-02-08-04-46-58.bpo-35936.q2ArjU.rst]: I think that makes sense, especially if the `imp` fix is eventually backed out.

**Comment by brandtbucher** [comment]
I’ve tweaked the implementation to not need a `fullname`, avoiding the issues mentioned above. I am still getting (what appear to be) random test failures, I suspect somehow stemming from my modifications to `sys.path`. This weird behavior is even referenced in an old comment in `test_modulefinder.py`, and I’m not able to reproduce the failures on my own machine.

Unless someone sees a clear fix for this odd behavior, I think I’ll just revert the new `importlib` bit of this PR. Bummer.

**Comment by brandtbucher** [comment]
Alright, I've gotten all of the tests passing for the new `importlib` implementation.

It turns out that the cached importers in `sys.path_importer_cache` were causing trouble, especially since so many of our tests use the same module names. My workaround is to copy and reset the _entire_ import state (`sys.path`, `sys.modules`, and `sys.path_importer_cache`) before each search, and swap the old values back in after. A  nice benefit of this approach is that path modifications work properly, and there is no longer a risk of accidentally importing a parent package when searching for a child module.

I broke these modifications out into a new private function, `_find_module`, so that it's clear how this fix is a simple drop-in replacement for `imp.find_module`. Let me know if there's anything I've missed or could improve. Otherwise, I feel my work here is done!

**Comment by vstinner** [comment]
@brettcannon @ncoghlan @warsaw: Would you mind to review this PR?

**Comment by pablogsal** [comment]
Why there are 3 NEWS entry for this PR?

**Comment by brandtbucher** [comment]
@pablogsal see @pganssle’s resolved comment above. I originally only had one, but they noted that this PR might benefit from more since it covers a few related (but separate) issues.

**Comment by vstinner** [comment]
As @pablogsal, I would also *prefer* a single NEWS entry ;-)

**Comment by pganssle** [comment]
The original NEWS entry made reference to "several long standing bugs", I assumed they had their own bpo entries.

That said if the PR fixes three things it's not uncommon in other projects to use three separate news entries. I haven't looked closely at the code of the PR, but the text of the NEWS entries seems to indicate this this fixing three mostly unrelated things.

File: Lib/modulefinder.py
Comment (ncoghlan) [inline Lib/modulefinder.py]: While this is a true statement, it's also substantial scope creep for `modulefinder`, which has historically only worked for regular filesystem paths. So I think it would be good to break this aspect of the change up into two parts:

1. In this PR, use [importlib.machiner.PathFinder.find_spec](https://docs.python.org/3/library/importlib.html#importlib.machinery.PathFinder) *without* clearing any of the import system caches to emulate the `imp.find_module` behaviour

2. File a follow-up enhancement request to expand module finder to respect `sys.meta_path` hooks

**Comment by ncoghlan** [comment]
As @pganssle notes - it makes sense to have multiple NEWS entries, since the issue and PR are a "roll-up" PR that allows a few different changes to be reviewed and tested together.

However, the news entries should be tweaked so their file names reference the more specific bug reports, rather than the summary issue.

File: Lib/modulefinder.py
Comment (ncoghlan) [reply Lib/modulefinder.py]: I realised I left out why the change concerns me and I'd like to see it reviewed separately: the actual `imp.find_module` implementation doesn't have any side effects, so calling it from multiple threads is safe. Rebinding `sys` module attributes on the other hand is decidedly *un*safe, so if that enhancement is going to happen, we'll need to find a different way of doing it that doesn't have global side effects :)

Comment (brandtbucher) [reply Lib/modulefinder.py]: Thanks for the feedback @ncoghlan. I've replaced the use of `importlib.util.find_spec` with `importlib.machinery.PathFinder.find_spec` for now.

On concurrency - I agree that it is a valid concern here. It would be trivial to include the proposed changes inside of a `threading.Lock` if we keep the `sys` modifications, but I can break that all out into a separate issue + PR if you feel it needs more discussion/review. I personally see `sys.meta_path`-respecting behavior as more of a "fix" than an "enhancement", but I'll defer to your judgement here.

**Comment by brandtbucher** [comment]
I've gone ahead with @ncoghlan's suggestion and updated two of the `NEWS` entries with filenames that reference their corresponding BPO issue numbers.

I've also updated the `_find_module` implementation to use `importlib.machinery.PathFinder.find_spec`. With that said, I feel that the existing scope of this PR justifies replacing this with my suggested `threading.Lock`/`sys`/`importlib.util.finc_spec` fix, rather than breaking out that change into a separate issue.

Any other thoughts?

File: Lib/modulefinder.py
Comment (ncoghlan) [inline Lib/modulefinder.py:128]: While this seems like a good change to me, it also seems worth mentioning in one of the NEWS entries (probably by adding it to https://github.com/python/cpython/pull/11787/files#diff-0c343f3267eee48d7cb592115947061b ).

**Review by ncoghlan — COMMENTED** [review]
One minor release docs comment, but otherwise this version looks good to me. However, I'd really like @ronaldoussoren to take another look before we merge it.

File: Lib/modulefinder.py
Comment (ronaldoussoren) [inline Lib/modulefinder.py]: ``(path+sys.path)`` changes the semantics that existed before this patch (_find_module might find modules on ``sys.path`` that aren't found on ``path``).

Why is it necessary to include ``sys.path`` on the search path here?

File: Lib/modulefinder.py
Comment (ronaldoussoren) [inline Lib/modulefinder.py:61]: I'd add a comment here that explains why clearing the cache is necessary, the need is not immediately obvious.

File: Lib/modulefinder.py
Comment (ronaldoussoren) [inline Lib/modulefinder.py]: Not sure if using ``issubclass`` necessary, just testing for these two special cases is good enough:

```
if spec.loader is import lib.machinery.BuiltinImporter:
  ...
```

Another special case is the loader for implicit namespace packages (which is currently an instance of a private class, see bpo-35673). Those are probably not handled at all at the moment by this module.

File: Lib/modulefinder.py
Comment (ronaldoussoren) [inline Lib/modulefinder.py]: I wonder if this behaviour should be emulated, although there's bound to be someone that relies on the current behaviour.

File: Lib/modulefinder.py
Comment (ronaldoussoren) [inline Lib/modulefinder.py:84]: I'm not sure about the code that follows. The *kind* can be deduced from the type of loader (at least for loaders in the stdlib).

File: Lib/modulefinder.py
Comment (ronaldoussoren) [inline Lib/modulefinder.py]: - Loader has a method ``is_package()`` that does this test. 

- The package ``__init__`` can also be an extension (although I'm not aware of any code that actually does this)

File: Lib/modulefinder.py
Comment (brandtbucher) [reply Lib/modulefinder.py]: Appending `sys.path` is left over from my early work on this change.. and I don't actually remember what led me to add it there. Either way, I agree that it is no longer necessary.

File: Lib/modulefinder.py
Comment (brandtbucher) [reply Lib/modulefinder.py]: Good news: it looks like `PathFinder` does this weird path formatting for us! So either way, no need for this ugly code.

File: Lib/modulefinder.py
Comment (brandtbucher) [reply Lib/modulefinder.py:84]: I've replaced extension checking with `isinstance` checks against the concrete loaders in `importlib.machinery`. I agree that it appears clearer and more robust than `imp`-style suffix checks.

File: Lib/modulefinder.py
Comment (brandtbucher) [reply Lib/modulefinder.py]: Thanks, didn't know about `is_package`. That makes it easy to move this check out in front of this branch.

File: Lib/modulefinder.py
Comment (brandtbucher) [reply Lib/modulefinder.py]: I've refactored these to use identity checking, as you suggested. Much cleaner.

I think it makes the most sense to wait on bpo-35673's resolution before trying to hack our way around namespace packages here... that could be a separate PR. What are your thoughts @ronaldoussoren?

File: Lib/modulefinder.py
Comment (brandtbucher) [reply Lib/modulefinder.py:61]: Thanks. I've clarified this bit in the source.

**Comment by brandtbucher** [comment]
Thanks @ronaldoussoren for the suggestions... everything is much cleaner now!

**Comment by brandtbucher** [comment]
*(bump)*

**Review by ncoghlan — APPROVED** [review]
Same situation as before - while I'm personally happy with the way the importlib APIs are now being used, @ronaldoussoren is the main reviewer from the actual `modulefinder` functionality side.

**Comment by brandtbucher** [comment]
@ronaldoussoren, do these changes look good to you?

**Comment by brandtbucher** [comment]
It's been six weeks since @ronaldoussoren last requested changes. Should I keep waiting for a response from him, or can this be merged?

**Comment by ncoghlan** [comment]
@brandtbucher Aye, I believe you've properly addressed all of @ronaldoussoren's feedback, so I've gone ahead and merged it since it appears he isn't currently available for a final review. Thanks for the update!

@ronaldoussoren If you spot anything further that I missed, just add a new comment and we can look at doing a further follow-up PR.

**Pull request merged**
